File: 743.network-delay-time.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.PriorityQueue;

/*
 * @lc app=leetcode id=743 lang=java
 *
 * [743] Network Delay Time
 */

// @lc code=start
class Solution {

    class Cell implements Comparable<Cell> {
        int node, time;
        Cell(int node, int time) {
            this.node = node;
            this.time = time;
        }    
        public int compareTo(Cell c2) {
            return  time-c2.time;
        }
    }



    public int networkDelayTime(int[][] times, int n, int k) {
     HashMap<Integer,List<Cell>> graph = new HashMap();
     boolean[] visited = new boolean[n+1];
     for (int[] edge : times) {
         graph.computeIfAbsent(edge[0], v->new ArrayList<>()).add( new Cell(edge[1],edge[2]));
     }
     Map<Integer, Integer> costs = new HashMap<>();
     PriorityQueue<Cell> heap = new PriorityQueue<Cell>();
     heap.add(new Cell(k, 0));
        
     while(!heap.isEmpty()) {
         Cell curr = heap.poll();
         if(costs.containsKey(curr.node)) continue;
         costs.put(curr.node, curr.time);
         
         if (graph.containsKey(curr.node))  {
             for(Cell neig: graph.get(curr.node)) {
                 if (!costs.containsKey(neig.node)) {
                     heap.add(new Cell(neig.node, neig.time+curr.time));
                 }
             }
         }
     }
     int ans = -1;
     if (costs.size() != n) return ans;
     for (int cost: costs.values()) {
         ans  = Math.max(ans, cost);
     }
     return ans;     
        
    }
}
// @lc code=end

File: 743.network-delay-time.py
#
# @lc app=leetcode id=743 lang=python3
#
# [743] Network Delay Time
#

# @lc code=start
from queue import PriorityQueue
from typing import List, Tuple


class Solution:
    
    def make_adjacency_list(self, arr:List[List[int]], n:int)->List[List[Tuple[int, int]]]:
        self.adj = [[] for _ in range(n+1)]
        for relation in arr:
            print(f"Error at: {relation} ")
            self.adj[relation[0]].append(( relation[2],relation[1]))
        return self.adj
                
    def dijkstra(self, n:int, k:int)->int:
        distance = [float('inf')]*(n+1)
        visited = [False]*(n+1)
        distance[k]=0
        queue = PriorityQueue()
        queue.put((0,k))
        while queue:
            current:Tuple[int,int] = queue.get()
            visited[current[1]]=True
            while self.adj[current[1]]:
                next_node:Tuple[int,int] = self.adj[current[1]].pop(0)
                if not visited[next_node[1]]:
                    distance[next_node[1]] = min(distance[next_node[1]], distance[current[1]]+next_node[0])
                    queue.put((distance[next_node[1]], next_node[1]))
        return max(distance) if max(distance) != float('inf') else -1
            
            
        
        
    
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        self.make_adjacency_list(times, n)
        return self.dijkstra(n, k)
        

        
# @lc code=end

